File: Data Structures & Algorithms/minimum-stack/submission-4.java
class MinStack {

    ArrayList<Integer> minStack = new ArrayList<>();
    ArrayList<Integer> stack = new ArrayList<>();
    
    public void push(int val) {
        
        if (minStack.isEmpty() || minStack.get(minStack.size() - 1) >= val){
            minStack.add(val);
        }

        stack.add(val);
    }
    
    public void pop() {
        if(minStack.get(minStack.size() - 1).equals(stack.get(stack.size() - 1))){
            minStack.remove(minStack.size() - 1);
        }

        stack.remove(stack.size() - 1);
    }
    
    public int top() {
        return stack.get(stack.size() - 1);
    }
    
    public int getMin() {
        return minStack.get(minStack.size() - 1);
    }
}

File: Data Structures & Algorithms/baseball-game/submission-4.py
class Solution:
    def calPoints(self, operations: List[str]) -> int:
        # Empty Record
        # Apply to the record operations 
            # Integer: record a new score
            # + - record a new score that is the sum of the previous two scores
            # 'D' - Record a new score double the previous one 
            # 'C' - Remove the record

        stack = []
        for action in operations:
            if action == "+":
                one = stack[-1]
                two = stack[-2]
                stack.append(one + two)
            elif action == "D":
                stack.append(stack[-1] * 2)
            elif action == "C":
                stack.pop()
            else:
                stack.append(int(action))
        
        sol = 0 
        while stack:
            sol += stack.pop()
        
        return sol
